src/Evolution/Systems/ScalarWave/UpwindPenaltyCorrection.hpp

```diff
@@ -0,0 +1,225 @@
+ * \f}
+ *
+ * Note that Eq. (6.3) of \cite Teukolsky2015ega is not exactly what's
+ * implemented since that boundary term does not consistently treat both sides
+ * of the interface on the same footing.
+ *
+ * For the scalar wave system the correction is:
+ *
+ * \f{align}{
+ *   D_{\Psi} &= \tilde{\lambda}_{v^{\Psi}}^{\mathrm{ext}}
+ *                v^{\mathrm{ext},\Psi}
+ *                - \tilde{\lambda}_{v^{\Psi}}^{\mathrm{int}}
+ *                v^{\mathrm{int},\Psi}, \\
+ *   D_{\Pi} &= \frac{1}{2}\left(\tilde{\lambda}_{v^+}^{\mathrm{ext}}
+ *             v^{\mathrm{ext},+} +
+ *             \tilde{\lambda}_{v^-}^{\mathrm{ext}}
+ *             v^{\mathrm{ext},-}\right)
+ *             + \tilde{\lambda}_{v^\Psi}^\mathrm{ext}\gamma_2
+ *             v^{\mathrm{ext},\Psi}
+ *             \notag \\
+ *           &-\frac{1}{2}\left(\tilde{\lambda}_{v^+}^{\mathrm{int}}
+ *             v^{\mathrm{int},+} +
+ *             \tilde{\lambda}_{v^-}^{\mathrm{int}}
+ *             v^{\mathrm{int},-}\right)
+ *             - \tilde{\lambda}_{v^\Psi}^\mathrm{int}\gamma_2
+ *             v^{\mathrm{int},\Psi} , \\
+ *   D_{\Phi_{i}}
+ *              &= \frac{1}{2}\left(\tilde{\lambda}_{v^+}^{\mathrm{ext}}
+ *                v^{\mathrm{ext},+}
+ *                - \tilde{\lambda}_{v^-}^{\mathrm{ext}}
+ *                v^{\mathrm{ext},-}\right)n_i^{\mathrm{ext}}
+ *                + \tilde{\lambda}_{v^0}^{\mathrm{ext}}
+ *                v^{\mathrm{ext},0}_{i}
+ *                \notag \\
+ *              &-
+ *                \frac{1}{2}\left(\tilde{\lambda}_{v^+}^{\mathrm{int}}
+ *                v^{\mathrm{int},+}
+ *                - \tilde{\lambda}_{v^-}^{\mathrm{int}}
+ *                v^{\mathrm{int},-}\right)n_i^{\mathrm{int}}
+ *                - \tilde{\lambda}_{v^0}^{\mathrm{int}}
+ *                v^{\mathrm{int},0}_{i},
+ * \f}
+ *
+ * with characteristic fields
+ *
+ * \f{align}{
+ *   v^{\Psi} &= \Psi, \\
+ *   v^{0}_{i} &= (\delta^k_i-n^k n_i)\Phi_{kab}, \\
+ *   v^{\pm} &= \Pi\pm n^i\Phi_{i} -\gamma_2 \Psi,
+ * \f}
+ *
+ * and characteristic speeds
+ *
+ * \f{align}{
+ *   \lambda_{v^\Psi} =& -v^i_g n_i, \\
+ *   \lambda_{v^0} =& -v^i_g n_i, \\
+ *   \lambda_{v^\pm} =& \pm 1 - v^i_g n_i,
+ * \f}
+ *
+ * where \f$v_g^i\f$ is the mesh velocity. We have also defined
+ *
+ * \f{align}{
+ *   \tilde{\lambda}_{\hat{\alpha}} =
+ *    \left\{
+ *        \begin{array}{ll}
+ *          \lambda_{\hat{\alpha}} &
+ *            \mathrm{if}\;\lambda_{\hat{\alpha}}\le 0.0 \\
+ *          0 & \mathrm{otherwise}
+ *        \end{array}\right.
+ * \f}
+ *
+ * Note that we have assumed \f$n_i^{\mathrm{ext}}\f$ points in the same
+ * direction as \f$n_i^{\mathrm{int}}\f$. If \f$n_i^{\mathrm{ext}}\f$ points in
+ * the opposite direction the external speeds have their sign flipped and the
+ * \f$\pm\f$ fields and speeds reverse roles.
+ */
+template <size_t Dim>
+struct UpwindPenaltyCorrection : tt::ConformsTo<dg::protocols::NumericalFlux> {
```

Comment:
I'm sure I'm just not seeing it, can you please help me understand how this replaces `PenaltyFlux` in addition to `Upwind`?

Reply:
I actually don't understand where the upwind as coded in develop comes from or why it's stable. In particular, I don't understand why there is more than one `F^i n_i` in the G for Phi_i. I'm not sure if I'm just not able to do math right now, but it looks like the upwind flux is almost the same as an upwind multipenalty modulo some signs that I can't get to work out right now. This makes me skeptical about how upwind it actually is. One reason I am removing it is because I don't trust that it actually is based on something solid mathematically since having different fluxes in the RHS of the G for phi is nonsense. Sorry I can't give you a better answer

Reply:
Okay, so having thought more about this, I think the upwind flux comes for Eq. 6.3 in Saul's paper. It turns out this is exactly the same as an upwind multipenalty with a penalty factor of 1 (what we use in DG). So the reason this can replace both is that the two coded up are the same boundary terms.

Reply:
I don't fully understand upwind as coded in develop, either...it was just the first thing I was able to get working. Are you saying you understand how UpwindFlux is related to UpwindPenlatyCorrection now, @nilsdeppe? Are you saying they are actually equivalent?

Reply:
@geoffrey4444 the nice thing is this applies to GH, SW, CSW. Yes, I now understand what PenaltyFlux, UpwindFlux are doing. UpwindFlux is, as far as I can tell, doing Eq. 6.3 in Saul's DG paper. The PenaltyFlux only penalizes ingoing char fields, and thus is the same as the UpwindFlux. If you look at the zero grid velocity limit of the UpwindPenaltyCorrection you see that you get back the PenaltyFlux, so the reason why UpwindPenaltyCorrection actually replaces both UpwindFlux and PenaltyFlux is because those two are equivalent (if the normals were treated equivalently).

Reply:
That still leaves the question about the name: I now have actually found a reason for the name UpwindPenaltyCorrection, it is an upwind penalty scheme, where as one could also do a simple penalty scheme where both the upwind and downwind cases are penalized.